lib/cache.js

```diff
@@ -131,6 +201,25 @@ with --force.`)
     this.npm.output(`Index entries: ${stats.totalEntries}`)
     this.npm.output(`Finished in ${stats.runTime.total / 1000}s`)
   }
+
+  // npm cache ls [--package <spec> ...]
+  async ls (args) {
+    const cachePath = path.join(this.npm.cache, '_cacache')
+    const cacheKeys = Object.keys(await cacache.ls(cachePath))
+    if (Array.isArray(this.npm.flatOptions.package)
+      && this.npm.flatOptions.package.length > 0) {
+      // get results for each package spec specified
+      const results = new Set()
+      for (const spec of this.npm.flatOptions.package) {
+        const keySet = await searchCachePackage(cachePath, spec, cacheKeys)
+        for (const key of keySet)
+          results.add(key)
+      }
+      [...results].sort().forEach(key => this.npm.output(key))
```

Comment:
just want to confirm: are we using a `Set` for results (and for the `cacheKeys`, which is converted to a `Set` in the `searchCachePackage` function) to avoid duplicates? do we really get duplicates?

also, when sorting package names elsewhere we generally use `localeCompare`, it feels like it would be best to follow suit here and change this to something like:

```suggestion
      [...results].sort((a, b) => a.localeCompare(b, 'en')).forEach((key) => this.npm.output(key))
```

Reply:
> just want to confirm: are we using a Set for results (and for the cacheKeys, which is converted to a Set in the searchCachePackage function) to avoid duplicates? do we really get duplicates?

answered elsewhere, the duplicates stem from us searching both for the specific urls listed in the packument as well as scanning all keys for the known URL pattern. we end up finding each key for a tarball in the public registry twice